staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go

```diff
@@ -0,0 +1,279 @@
+		result := &cachingObject{object: obj.DeepCopyObject().(metaRuntimeInterface)}
+		result.serializations.Store(make(serializationsCache))
+		return result, nil
+	}
+	return nil, fmt.Errorf("can't cast object to metav1.Object: %#v", object)
+}
+
+// CacheEncode implements runtime.CacheableObject interface.
+// It serializes the object and writes the result to given io.Writer trying
+// to first use the already cached result and falls back to a given encode
+// function in case of cache miss.
+// It assumes that for a given identifier, the encode function always encodes
+// each input object into the same output format.
+func (o *cachingObject) CacheEncode(id runtime.Identifier, encode func(runtime.Object, io.Writer) error, w io.Writer) error {
+	result := func() *serializationResult {
+		// Fast-path for getting from cache.
+		serializations := o.serializations.Load().(serializationsCache)
+		if result, exists := serializations[id]; exists {
+			return result
+		}
+
+		// Slow-path (that may require insert).
+		o.lock.Lock()
+		defer o.lock.Unlock()
+
+		serializations = o.serializations.Load().(serializationsCache)
+		// Check if in the meantime it wasn't inserted.
+		if result, exists := serializations[id]; exists {
+			return result
+		}
+
+		// Insert an entry for <id>. This requires copy of existing map.
+		newSerializations := make(serializationsCache)
+		for k, v := range serializations {
+			newSerializations[k] = v
+		}
+		result := &serializationResult{}
+		newSerializations[id] = result
+		o.serializations.Store(newSerializations)
+		return result
+	}()
+	result.once.Do(func() {
+		buffer := bytes.NewBuffer(nil)
+		result.err = encode(o.GetObject(), buffer)
+		result.raw = buffer.Bytes()
+	})
+	// Once invoked, fields of serialization will not change.
+	if result.err != nil {
+		return result.err
+	}
+	_, err := w.Write(result.raw)
+	return err
+}
+
+// GetObject implements runtime.CacheableObject interface.
+func (o *cachingObject) GetObject() runtime.Object {
+	result := o.object.DeepCopyObject().(metaRuntimeInterface)
+	func() {
+		o.lock.Lock()
+		defer o.lock.Unlock()
+		if o.selfLink != nil {
+			result.SetSelfLink(*o.selfLink)
+		}
+	}()
+	return result
+}
+
+// GetObjectKind implements runtime.Object interface.
+func (o *cachingObject) GetObjectKind() schema.ObjectKind {
+	return o.object.GetObjectKind()
+}
+
+// DeepCopyObject implements runtime.Object interface.
+func (o *cachingObject) DeepCopyObject() runtime.Object {
+	// DeepCopyObject on cachingObject is not expected to be called anywhere.
+	// However, to be on the safe-side, we implement it, though given the
+	// cache is only an optimization we ignore copying it.
+	result := &cachingObject{}
+	result.object = o.object.DeepCopyObject().(metaRuntimeInterface)
+	result.serializations.Store(make(serializationsCache))
+	func() {
+		o.lock.Lock()
+		defer o.lock.Unlock()
+		if o.selfLink != nil {
+			selfLink := *o.selfLink
+			result.selfLink = &selfLink
```

Comment:
special-casing selfLink seems problematic. Can we generalize the object ownership principles so that the cached encoding can safely operate without this special-casing?

Reply:
Once we will remove selflink, this special case will also be removed - this is sth that I had in the back of my head too.

Reply:
can drop this as part of passing through setters

Reply:
done